architecture/2024-05-radius-on-dapr.md

```diff
@@ -0,0 +1,381 @@
+
+## Compatibility (optional)
+
+We don't currently support upgrades from one release to another. Adopting Dapr and mature database technology like Postgres will help us support upgrades in the future.
+
+## Monitoring and Logging
+
+We should use our existing metrics and otel configuration settings to configure [Dapr's telemetry features](https://docs.dapr.io/operations/configuration/configuration-overview/#tracing).
+
+## Development plan
+
+Stage 0:
+
+- Build support for local development of Radius based on Dapr (off by default). This gives Radius the ability to launch its own Dapr sidecars for testing.
+- Build database layer for Radius based on Dapr statestore (off by default).
+- Update UCP integration tests to use Dapr statestore instead of etcd.
+- Build support for automatic component generation for UCP. 
+  - Acceptance criteria: UCP can complete its startup sequence in the debugger with no Kubernetes, only Dapr.
+
+*These work items is being frontloaded to increase productivity of the remaining work. When this is done will have good workflows for development and for integration tests.*
+
+Stage 1:
+
+- Add Dapr installation to `rad init` and `rad install`.
+- Add Dapr annotations to our Kubernetes components.
+- Add default components to our Helm chart.
+- Add startup-time verification to our services to verify correct configuration of Dapr and its components.
+
+Stage 2: 
+
+- Switch our Go-based services to use the Dapr state store.
+- Remove etcd-based, crd-based, and cosmosdb-based database implementations.
+
+Stage 3:
+
+- Implement new async-worker.
+  - Acceptance criteria: works in UCP integration tests with no Kubernetes, only Dapr.
+- Switch our Go-based services to use the new async-worker.
+- Remove queue abstraction and implementations from Radius.
+
+Stage 4:
+
+- Add Dapr sidecar management and startup-time verification to DE.
+- Replace DE in-memory data layer with Dapr state store, update integration tests.
+
+## Open Questions / FAQ
+
+### Decisions
+
+*These are topics in-scope of this document, where we've had enough discussion to make a decision.*
+
+**Q: Can we use the Dapr state store as a database? What are the tradeoffs?**
+
+The Dapr state store building block provides the capabilities of a generic key-value-store. This is lower level than a document store like MongoDB or a RDBMS like PostgreSQL. 
+
+This isn't bad, we designed Radius around a key-value-store abstraction. Our existing database abstraction is similar to Dapr's state store. It's a good fit.
+
+There's one minor issue to address which is how to implement "list" operations (eg: list all containers in resource group X). Dapr state store doesn't have list functionality (though most key-value-stores do). The well-known solution for this is to maintain our own "index" as it's own key. This is a common enough solution that we shouldn't be concerned. 
```

Comment:
Do we have this functionality right now? 🧐 I thought we didn't. Both ways, it could be a good challenge to solve.

Reply:
Yes we have this functionality now.